### PR TITLE
Link statically with custom jemalloc with disabled initial TLS

### PR DESCRIPTION
**Pull request by ldorau**
### Description

Link statically with custom jemalloc built from sources
with the following non-default options enabled:
- `--with-jemalloc-prefix=je_` - add je_ prefix to all public APIs
- `--disable-cxx` - Disable C++ integration. This will cause new and delete operators implementations to be omitted.
- `--disable-initial-exec-tls` - Disable the initial-exec TLS model for jemalloc's internal thread-local storage (on those platforms that support explicit settings). This can allow jemalloc to be dynamically loaded after program startup (e.g. using `dlopen()`).

Fixes: #891
Fixes: #894
Fixes: #903

### Checklist

- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

### CODE REVIEW

**Comment by ldorau** [comment]
Fixes: #891
Fixes: #894

File: examples/dram_and_fsdax/CMakeLists.txt [hunk 1]
```diff
@@ -21,10 +21,8 @@ if(NOT LIBHWLOC_FOUND)
     find_package(LIBHWLOC 2.3.0 REQUIRED hwloc)
 endif()
 
-pkg_check_modules(JEMALLOC jemalloc)
-if(NOT JEMALLOC_FOUND)
-    find_package(JEMALLOC REQUIRED jemalloc)
-endif()
+# find custom jemalloc pointed by CMAKE_PREFIX_PATH (with je_ prefix enabled)
```
[the remark this answers is not in the document]
Comment (ldorau) [reply examples/dram_and_fsdax/CMakeLists.txt]: Done.

File: .github/workflows/reusable_multi_numa.yml [hunk 2]
```diff
@@ -45,8 +46,9 @@ jobs:
           -DUMF_BUILD_TESTS=ON
           -DUMF_DEVELOPER_MODE=ON
           -DUMF_BUILD_LIBUMF_POOL_DISJOINT=ON
-          -DUMF_BUILD_LIBUMF_POOL_JEMALLOC=ON
           -DUMF_TESTS_FAIL_ON_SKIP=ON
+          ${{ matrix.os == 'ubuntu-22.04' && '-DUMF_BUILD_LIBUMF_POOL_JEMALLOC=ON' || '' }}
```
[the remark this answers is not in the document]
Comment (ldorau) [reply .github/workflows/reusable_multi_numa.yml]: Done

File: .github/workflows/reusable_multi_numa.yml [hunk 3]
```diff
@@ -33,6 +33,7 @@ jobs:
       - name: Get information about platform
         run: .github/scripts/get_system_info.sh
 
+      # jemalloc pool on RHEL is disabled, because the autoconf package cannot be installed, since the RHEL subscription expired
```
[the remark this answers is not in the document]
Comment (ldorau) [reply .github/workflows/reusable_multi_numa.yml]: Removed.
Done.

**Comment by ldorau** [comment]
There is still one issue:
- `./benchmark/umf-bench-ubench` segfaults for unknown reason **on L0 only** (the segfault occurs in jemalloc's `je_free_default()` called from the constructor of `libigc.so` which `libze_intel_gpu.so.1` depends on) see: https://github.com/oneapi-src/unified-memory-framework/actions/runs/12069615645/job/33659028476

**Comment by ldorau** [comment]
> There is still one issue:
> 
> * `./benchmark/umf-bench-ubench` segfaults for unknown reason **on L0 only** (the segfault occurs in jemalloc's `je_free_default()` called from the constructor of `libigc.so` which `libze_intel_gpu.so.1` depends on) see: https://github.com/oneapi-src/unified-memory-framework/actions/runs/12069615645/job/33659028476

I am testing the previous release of jemalloc (v5.2.1) now...

**Comment by ldorau** [comment]
@bratpiorka changing prefix to je_umf_ did not fix the issue ... :-(

**Comment by ldorau** [comment]
@bratpiorka @lplewa review please

**Review by bratpiorka — APPROVED** [review]

File: CMakeLists.txt [hunk 4]
```diff
@@ -111,6 +111,87 @@ else()
     message(FATAL_ERROR "Unknown OS type")
 endif()
 
+if(NOT UMF_BUILD_LIBUMF_POOL_JEMALLOC)
+    message(STATUS "jemalloc is disabled")
```
Comment (lukaszstolarczuk) [inline CMakeLists.txt:115]: I guess we don't have any such msg when an option is disabled

File: .github/docker/ubuntu-20.04.Dockerfile [hunk 5]
```diff
@@ -24,7 +24,6 @@ ARG BASE_DEPS="\
 
 # UMF's dependencies
 ARG UMF_DEPS="\
```
Comment (lukaszstolarczuk) [inline .github/docker/ubuntu-20.04.Dockerfile:26]: I know we're not testing these dockers currently, but perhaps add `automake` to `MISC_DEPS` (both dockers)

**Review by lukaszstolarczuk — COMMENTED** [review]
in general, LGTM

File: README.md [hunk 6]
```diff
@@ -262,7 +262,7 @@ The `UMF_BUILD_LIBUMF_POOL_JEMALLOC` option has to be turned `ON` to build this
 
 1) The `UMF_BUILD_LIBUMF_POOL_JEMALLOC` option turned `ON`
 2) Required packages:
-   - libjemalloc-dev (Linux) or jemalloc (Windows)
+   - jemalloc (Windows only)
```
[the remark this answers is not in the document]
Comment (ldorau) [reply README.md:265]: Added to `README.md`

File: CMakeLists.txt (hunk 4 above)
Comment (ldorau) [reply CMakeLists.txt:115]: Replaced with `set(UMF_POOL_JEMALLOC_ENABLED FALSE)`

File: .github/docker/ubuntu-20.04.Dockerfile (hunk 5 above)
Comment (ldorau) [reply .github/docker/ubuntu-20.04.Dockerfile:26]: Done

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply README.md:279]: thx!

**Comment by ldorau** [comment]
@vinser52 @lplewa @PatKamin @kswiecicki @KFilipek please review

File: examples/CMakeLists.txt [hunk 7]
```diff
@@ -284,8 +284,11 @@ if(LINUX)
             SRCS dram_and_fsdax/dram_and_fsdax.c
             LIBS umf jemalloc_pool)
```
Comment (vinser52) [inline examples/CMakeLists.txt:285]: Do we still have `jemalloc_pool` library? Now jemalloc statically linked to the UMF right? So, should example still link with `jemalloc_pool`?

File: examples/dram_and_fsdax/CMakeLists.txt [hunk 8]
```diff
@@ -21,10 +21,8 @@ if(NOT LIBHWLOC_FOUND)
     find_package(LIBHWLOC 2.3.0 REQUIRED hwloc)
 endif()
 
-pkg_check_modules(JEMALLOC jemalloc)
-if(NOT JEMALLOC_FOUND)
-    find_package(JEMALLOC REQUIRED jemalloc)
-endif()
+# find the custom jemalloc pointed by CMAKE_PREFIX_PATH
```
Comment (vinser52) [inline examples/dram_and_fsdax/CMakeLists.txt:24]: What if there is no custom `jemalloc` library on the system?

File: src/pool/CMakeLists.txt [hunk 9]
```diff
@@ -45,10 +45,13 @@ if(UMF_BUILD_LIBUMF_POOL_JEMALLOC)
...
         TYPE STATIC
         SRCS pool_jemalloc.c ${POOL_EXTRA_SRCS}
-        LIBS jemalloc ${POOL_EXTRA_LIBS})
+        LIBS ${JEMALLOC_LIBRARIES} ${POOL_EXTRA_LIBS})
     target_include_directories(jemalloc_pool PRIVATE ${JEMALLOC_INCLUDE_DIRS})
     target_compile_definitions(jemalloc_pool
                                PRIVATE ${POOL_COMPILE_DEFINITIONS})
     add_library(${PROJECT_NAME}::jemalloc_pool ALIAS jemalloc_pool)
+    if(NOT WINDOWS)
+        add_dependencies(jemalloc_pool jemalloc)
```
Comment (vinser52) [inline src/pool/CMakeLists.txt:54]: Why do we still have `jemalloc_pool`?

File: examples/CMakeLists.txt (hunk 7 above)
Comment (ldorau) [reply examples/CMakeLists.txt:285]: Yes, we still have the static `jemalloc_pool` library.
No, jemalloc is statically linked to the `jemalloc_pool` library (not to the UMF).
Yes, the example should still link with the `jemalloc_pool` library, because it uses its API (`umfJemallocPoolOps` for example).

Comment (vinser52) [reply examples/CMakeLists.txt:285]: Hmm, why did we make such a decision to keep the static `jemalloc_pool` library? Other pools are part of libumf. Why do we need a separate static `jemalloc_pool` library?

File: src/pool/CMakeLists.txt (hunk 9 above)
Comment (vinser52) [reply src/pool/CMakeLists.txt:54]: I am confused, do we have separate library for a `scalable_pool`? I thought no.
For a `disjoint_pool` we have a separate static library because today it is implemented in C++. Once @bratpiorka rewrites it in C it will be part of `libumf.so`.

File: examples/CMakeLists.txt (hunk 7 above)
Comment (ldorau) [reply examples/CMakeLists.txt:285]: OK, I understand now. I will fix it.

File: src/pool/CMakeLists.txt (hunk 9 above)
Comment (ldorau) [reply src/pool/CMakeLists.txt:54]: OK, I understand now. I will fix it.

File: CMakeLists.txt [hunk 10]
```diff
@@ -111,6 +111,87 @@ else()
...
+    set(JEMALLOC_LIBRARY_DIRS ${jemalloc_targ_BINARY_DIR}/lib)
+    set(JEMALLOC_INCLUDE_DIRS ${jemalloc_targ_BINARY_DIR}/include)
+    set(JEMALLOC_LIBRARIES ${jemalloc_targ_BINARY_DIR}/lib/libjemalloc.a)
+endif()
+
+if(JEMALLOC_FOUND OR JEMALLOC_LIBRARIES)
+    set(UMF_POOL_JEMALLOC_ENABLED TRUE)
+    # add PATH to DLL on Windows
+    set(DLL_PATH_LIST
+        "${DLL_PATH_LIST};PATH=path_list_append:${JEMALLOC_DLL_DIRS}")
```
Comment (PatKamin) [inline CMakeLists.txt:189]: Is it needed if a static lib is used now?

File: CMakeLists.txt [hunk 11]
```diff
@@ -111,6 +111,87 @@ else()
...
+        # --disable-cxx - Disable C++ integration. This will cause new and
+        # delete operators implementations to be omitted.
+        # --disable-initial-exec-tls - Disable the initial-exec TLS model for
+        # jemalloc's internal thread-local storage (on those platforms that
+        # support explicit settings). This can allow jemalloc to be dynamically
+        # loaded after program startup (e.g. using dlopen).
+        COMMAND
+            ./configure --prefix=${jemalloc_targ_BINARY_DIR}
+            --with-jemalloc-prefix=je_ --disable-cxx --disable-initial-exec-tls
+            CFLAGS=-fPIC CXXFLAGS=-fPIC
```
[the remark this answers is not in the document]
Comment (ldorau) [reply CMakeLists.txt]: Done.

File: src/pool/CMakeLists.txt (hunk 9 above)
Comment (ldorau) [reply src/pool/CMakeLists.txt]: Removed.
Done.

File: examples/dram_and_fsdax/CMakeLists.txt (hunk 8 above)
Comment (ldorau) [reply examples/dram_and_fsdax/CMakeLists.txt]: Removed.
Done.

File: examples/CMakeLists.txt (hunk 7 above)
Comment (ldorau) [reply examples/CMakeLists.txt]: No, removed.
Done.

File: .github/workflows/performance.yml [hunk 12]
```diff
@@ -78,7 +78,6 @@ jobs:
         -DUMF_BUILD_LEVEL_ZERO_PROVIDER=ON
         -DUMF_BUILD_CUDA_PROVIDER=ON
         -DUMF_BUILD_LIBUMF_POOL_DISJOINT=ON
-        -DUMF_BUILD_LIBUMF_POOL_JEMALLOC=ON
```
Comment (vinser52) [inline .github/workflows/performance.yml:81]: What if the build machine has not Internet access? If it is not possible to download jemalloc sources what happens?

Comment (ldorau) [reply .github/workflows/performance.yml:81]: A user can provide its own local repo using `UMF_JEMALLOC_REPO`

Comment (PatKamin) [reply .github/workflows/performance.yml:81]: That's a good point. I think we should warn in the fetching steps and continue the build in such a case with `UMF_POOL_JEMALLOC_ENABLED=0`.

**Review by vinser52 — APPROVED** [review]
Please fix the remaining comment

**Review by PatKamin — APPROVED** [review]

**Review by lukaszstolarczuk — CHANGES_REQUESTED** [review]

File: README.md [hunk 13]
```diff
@@ -101,7 +101,7 @@ List of options provided by CMake:
 | UMF_BUILD_LEVEL_ZERO_PROVIDER | Build Level Zero memory provider | ON/OFF | ON |
 | UMF_BUILD_CUDA_PROVIDER | Build CUDA memory provider | ON/OFF | ON |
 | UMF_BUILD_LIBUMF_POOL_DISJOINT | Build the libumf_pool_disjoint static library | ON/OFF | OFF |
-| UMF_BUILD_LIBUMF_POOL_JEMALLOC | Build the libumf_pool_jemalloc static library | ON/OFF | OFF |
+| UMF_BUILD_LIBUMF_POOL_JEMALLOC | Build the libumf_pool_jemalloc static library | ON/OFF | ON |
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply README.md:104]: I think @lukaszstolarczuk is right here. BTW this will change oneAPI toolkit where we currently do not have jemalloc. IMO it should be disabled by default

File: benchmark/multithread.cpp [hunk 14]
```diff
@@ -139,11 +139,13 @@ int main() {
     // ctest looks for "PASSED" in the output
     std::cout << "PASSED" << std::endl;
 
+#if defined(UMF_BUILD_LIBUMF_POOL_DISJOINT)
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply benchmark/multithread.cpp]: In our CMake files *_ENABLED = user set appropriate BUILD flag + all dependencies needed for build are satisfied so we actually can use given component. 
@ldorau please use *_ENABLED in c/cpp files

[the remark this answers is not in the document]
Comment (ldorau) [reply benchmark/multithread.cpp]: Done

File: .github/workflows/coverity.yml [hunk 15]
```diff
@@ -50,7 +50,6 @@ jobs:
         -DUMF_DEVELOPER_MODE=OFF
         -DUMF_TESTS_FAIL_ON_SKIP=ON
         -DUMF_BUILD_LIBUMF_POOL_DISJOINT=ON
```
Comment (bratpiorka) [inline .github/workflows/coverity.yml:52]: please revert this change and disable jemalloc by default here and in other places

File: CMakeLists.txt [hunk 16]
```diff
@@ -455,14 +531,14 @@ elseif(UMF_PROXY_LIB_BASED_ON_POOL STREQUAL SCALABLE)
...
+    if(UMF_POOL_JEMALLOC_ENABLED)
         set(UMF_PROXY_LIB_ENABLED ON)
         set(PROXY_LIB_USES_JEMALLOC_POOL ON)
-        set(PROXY_LIBS jemalloc_pool umf)
+        set(PROXY_LIBS umf)
     else()
         message(
             STATUS
-                "Disabling the proxy library, because UMF_PROXY_LIB_BASED_ON_POOL==JEMALLOC but UMF_BUILD_LIBUMF_POOL_JEMALLOC is OFF"
+                "Disabling the proxy library, because UMF_PROXY_LIB_BASED_ON_POOL==JEMALLOC but JEMALLOC was not found"
```
[the remark this answers is not in the document]
Comment (ldorau) [reply CMakeLists.txt]: Done

File: benchmark/CMakeLists.txt [hunk 17]
```diff
@@ -121,8 +121,7 @@ set(LIB_DIRS ${LIBHWLOC_LIBRARY_DIRS})
 if(UMF_BUILD_LIBUMF_POOL_DISJOINT)
     set(LIBS_OPTIONAL ${LIBS_OPTIONAL} disjoint_pool)
 endif()
-if(UMF_BUILD_LIBUMF_POOL_JEMALLOC)
-    set(LIBS_OPTIONAL ${LIBS_OPTIONAL} jemalloc_pool ${JEMALLOC_LIBRARIES})
+if(UMF_POOL_JEMALLOC_ENABLED)
     set(LIB_DIRS ${LIB_DIRS} ${JEMALLOC_LIBRARY_DIRS})
```
Comment (lukaszstolarczuk) [inline benchmark/CMakeLists.txt:125]: nit: so lib_dirs but no libs?

**Review by lukaszstolarczuk — APPROVED** [review]

File: benchmark/CMakeLists.txt (hunk 17 above)
Comment (ldorau) [reply benchmark/CMakeLists.txt:125]: Right, will remove in a separate PR

Comment (ldorau) [reply benchmark/CMakeLists.txt:125]: Done: https://github.com/oneapi-src/unified-memory-framework/pull/964
